src/write_gcm.c

```diff
@@ -3775,6 +3966,46 @@ static void wg_log_json_message(const wg_context_t *ctx, const char *fmt, ...) {
   fflush(ctx->json_log_file);
 }
 
+static char *wg_get_instance_id_from_monitored_resource(
+    monitored_resource_t *resource) {
+  if (resource == NULL) {
+    ERROR(
+        "write_gcm: wg_get_instance_id_from_monitored_resource resource empty");
+    return NULL;
+  }
+  for (int i = 0; i < resource->num_labels; i++) {
+    if (strcmp(resource->keys[i], "instance_id") == 0) {
+      return resource->values[i];
+    }
+  }
+  return NULL;
+}
+
+// Determines which Monitored Resource to use for the payload being dispatched.
+static int wg_determine_monitored_resource_for_payload(
+    const char *host, monitored_resource_t **response_ptr,
+    const wg_context_t *ctx) {
+  int result = -1;
+  monitored_resource_t *resource = NULL;
+  if (host != NULL && ctx->enable_metadata_agent) {
+    resource =
+        wg_monitored_resource_create_from_metadata_agent(
+            ctx->metadata_agent_url, ctx->resource->project_id, host);
+  }
+  if (resource != NULL) {
+    *response_ptr = resource;
+    result = 0;
+  } else {
+    char *instance_id =
```

Comment:
On a more general note - as the design has evolved since we decided to do this - should this logic be removed? i.e. if the metadata agent is unavailable just send all metrics using the default instance monitored resource? Or are we concerned about duplicate points in the same request causing the API to start dropping points?